nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -0,0 +1,765 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Nexus taking a snapshot is a perfect application for a saga. There are
+//! several services that requests must be sent to, and the snapshot can only be
+//! considered completed when all have returned OK and the appropriate database
+//! records have been created.
+//!
+//! # Taking a snapshot of a disk currently attached to an instance
+//!
+//! A running instance has a corresponding zone that is running a propolis
+//! server. That propolis server emulates a disk for the guest OS, and the
+//! backend for that virtual disk is a crucible volume that is constructed out
+//! of downstairs regions and optionally some sort of read only source (an
+//! image, another snapshot, etc). The downstairs regions are concatenated
+//! together under a "sub-volume" array over the span of the whole virtual disk
+//! (this allows for one 1TB virtual disk to be composed of several smaller
+//! downstairs regions for example). Below is an example of a volume where there
+//! are two regions (where each region is backed by three downstairs) and a read
+//! only parent that points to a URL:
+//!
+//! ```text
+//! Volume {
+//!   sub-volumes [
+//!      region A {
+//!        downstairs @ [fd00:1122:3344:0101::0500]:19000,
+//!        downstairs @ [fd00:1122:3344:0101::0501]:19000,
+//!        downstairs @ [fd00:1122:3344:0101::0502]:19000,
+//!      },
+//!      region A {
+//!        downstairs @ [fd00:1122:3344:0101::0503]:19000,
+//!        downstairs @ [fd00:1122:3344:0101::0504]:19000,
+//!        downstairs @ [fd00:1122:3344:0101::0505]:19000,
+//!      },
+//!   ]
+//!
+//!   read only parent {
+//!     image {
+//!       url: "some.url/debian-11.img"
+//!   }
+//! }
+//! ```
+//!
+//! When the guest OS writes to their disk, those blocks will be written to the
+//! sub-volume regions only - no writes or flushes are sent to the read only
+//! parent by the volume, only reads. A block read is served from the read only
+//! parent if there hasn't been a write to that block in the sub volume,
+//! otherwise the block is served from the sub volume. This means all modified
+//! blocks for a volume will be written to sub-volume regions, and it's those
+//! modifications we want to capture in a snapshot.
+//!
+//! First, this saga will send a snapshot request to the instance's propolis
+//! server. Currently a snapshot request is implemented as a flush with an extra
+//! parameter, and this is sent to the volume through the standard IO channels.
+//! This flush will be processed by each downstairs in the same job order so
+//! that each snapshot will contain the same information. A crucible snapshot is
```

Comment:
> Are operations to the downstairs totally ordered? I'm somewhat ignorant of the concurrency story there, but it seems like it would have a major impact on whether or not we'd expect these ZFS snapshots to be the same.

Operations processed by the downstairs are ordered by job ID, yeah.

> What happens if this operation is issued while one of the downstairs is offline / undergoing repair?

If one downstairs is offline or otherwise not healthy, then the flush will hit the other two downstairs and be considered acknowledged. It will be queued by the Upstairs and resent to the third downstairs when it is healthy (or I guess another one is swapped in).

If two downstairs are not healthy, then no writes will work. The saga will issue the snapshot request, then ask the crucible agent for the snapshot details - the agent will return that it can't find the snapshot, and the saga will unwind.